scm-plugins/scm-git-plugin/src/main/java/sonia/scm/repository/spi/GitHookChangesetCollector.java

```diff
@@ -0,0 +1,216 @@
+import org.eclipse.jgit.revwalk.RevWalk;
+import org.eclipse.jgit.transport.ReceiveCommand;
+import org.eclipse.jgit.transport.ReceivePack;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import sonia.scm.repository.Changeset;
+import sonia.scm.repository.GitChangesetConverter;
+import sonia.scm.repository.GitChangesetConverterFactory;
+import sonia.scm.repository.GitUtil;
+import sonia.scm.web.CollectingPackParserListener;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+
+import static java.util.Collections.unmodifiableCollection;
+
+/**
+ * @author Sebastian Sdorra
+ */
+class GitHookChangesetCollector {
+
+  private static final Logger LOG = LoggerFactory.getLogger(GitHookChangesetCollector.class);
+
+  /**
+   * listener to track new objects
+   */
+  private final CollectingPackParserListener listener;
+
+  private final Collection<ReceiveCommand> receiveCommands;
+
+  private final GitChangesetConverterFactory converterFactory;
+  private final ReceivePack rpack;
+
+  private final Map<String, Changeset> addedChangesets = Maps.newLinkedHashMap();
+  private final Map<String, Changeset> removedChangesets = Maps.newLinkedHashMap();
+
+  GitHookChangesetCollector(GitChangesetConverterFactory converterFactory, ReceivePack rpack, Collection<ReceiveCommand> receiveCommands) {
+    this.converterFactory = converterFactory;
+    this.rpack = rpack;
+    this.receiveCommands = receiveCommands;
+    this.listener = CollectingPackParserListener.get(rpack);
+  }
+
+  /**
+   * Collect all new changesets from the received hook. Afterwards, the results can be
+   * retrieved with {@link #getAddedChangesets()} and {@link #getRemovedChangesets()}
+   */
+  void collectChangesets() {
+    try (Repository repository = rpack.getRepository();
+         RevWalk walk = rpack.getRevWalk();
+         GitChangesetConverter converter = converterFactory.create(repository, walk)) {
+      repository.incrementOpen();
+
+      for (ReceiveCommand rc : receiveCommands) {
+        String ref = rc.getRefName();
+
+        LOG.trace("handle receive command, type={}, ref={}, result={}", rc.getType(), ref, rc.getResult());
+
+        handle(repository, walk, converter, rc, ref);
+      }
+    } catch (Exception ex) {
+      LOG.error("could not collect changesets", ex);
+    }
+  }
+
+  Iterable<Changeset> getAddedChangesets() {
+    return unmodifiableCollection(addedChangesets.values());
+  }
+
+  Iterable<Changeset> getRemovedChangesets() {
+    return unmodifiableCollection(removedChangesets.values());
+  }
+
+  void handle(Repository repository, RevWalk walk, GitChangesetConverter converter, ReceiveCommand rc, String ref) {
+    try {
+      if (!(GitUtil.isBranch(ref) || GitUtil.isTag(ref))) {
+        LOG.debug("skip ref {}, because it is neither branch nor tag", ref);
+      } else if (rc.getType() == ReceiveCommand.Type.UPDATE_NONFASTFORWARD) {
+        LOG.debug("handle deleted ref {}", ref);
```

Comment:
I think this log should state something like "overwritten/replaced", or am I looking at this wrong ?